src/roster/mod_roster.erl

```diff
@@ -316,58 +315,42 @@ create_sub_el(Items, Version) ->
-                                          true
-                                  end, get_roster(HostType, LUser, LServer)),
-            mongoose_acc:append(roster, items, Roster, Acc)
-    end.
-
+get_user_roster(Items, #{show_full_roster := Full, jid := JID}, #{host_type := HostType}) ->
+    NewItems = do_get_user_roster(HostType, JID, Full),
+    {ok, Items ++ NewItems}.
+
+-spec do_get_user_roster(mongooseim:host_type(), jid:jid(), boolean()) -> [roster()].
+do_get_user_roster(HostType, #jid{luser = LUser, lserver = LServer}, true) ->
+    get_roster(HostType, LUser, LServer);
+do_get_user_roster(HostType, #jid{luser = LUser, lserver = LServer}, false) ->
+    Roster = get_roster(HostType, LUser, LServer),
+    Fun = fun(#roster{subscription = S, ask = A}) -> not (none =:= S andalso in =:= A) end,
+    lists:filter(Fun, Roster).
+
+-spec item_to_xml(roster()) -> exml:element().
 item_to_xml(Item) ->
-    Attrs1 = [{<<"jid">>,
-               jid:to_binary(Item#roster.jid)}],
-    Attrs2 = case Item#roster.name of
-                 <<"">> -> Attrs1;
-                 Name -> [{<<"name">>, Name} | Attrs1]
-             end,
-    Attrs3 = case Item#roster.subscription of
-                 none -> [{<<"subscription">>, <<"none">>} | Attrs2];
-                 from -> [{<<"subscription">>, <<"from">>} | Attrs2];
-                 to -> [{<<"subscription">>, <<"to">>} | Attrs2];
-                 both -> [{<<"subscription">>, <<"both">>} | Attrs2];
-                 remove -> [{<<"subscription">>, <<"remove">>} | Attrs2]
-             end,
-    Attrs4 = case ask_to_pending(Item#roster.ask) of
-                 out -> [{<<"ask">>, <<"subscribe">>} | Attrs3];
-                 both -> [{<<"ask">>, <<"subscribe">>} | Attrs3];
-                 _ -> Attrs3
-             end,
-    SubEls1 = lists:map(fun (G) ->
-                                #xmlel{name = <<"group">>, attrs = [],
-                                       children = [{xmlcdata, G}]}
-                        end,
-                        Item#roster.groups),
-    SubEls = SubEls1 ++ Item#roster.xs,
-    #xmlel{name = <<"item">>, attrs = Attrs4,
-           children = SubEls}.
+    Attrs0 = [{<<"jid">>, jid:to_binary(Item#roster.jid)},
+              {<<"subscription">>, subs_to_binary(Item#roster.subscription)}],
+    Attrs = case {Item#roster.name, Item#roster.ask} of
+                {<<>>, Ask} when Ask =:= subscribe; Ask =:= out; Ask =:= both ->
+                    [{<<"ask">>, <<"subscribe">>} | Attrs0];
+                {Name, Ask} when Ask =:= subscribe; Ask =:= out; Ask =:= both ->
+                    [{<<"name">>, Name}, {<<"ask">>, <<"subscribe">>} | Attrs0];
+                {<<>>, _} ->
+                    Attrs0;
+                {Name, _} ->
+                    [{<<"name">>, Name} | Attrs0]
+            end,
+    Fold = fun(G, Acc) -> [group_el(G) | Acc] end,
+    SubEls = lists:foldl(Fold, Item#roster.xs, Item#roster.groups),
```

Comment:
I liked the fold only because it is faster to create a longer list by inserting at the beginning than it is to append two lists, but dunno how long the lists are or if it actually matters 🤔